### ISSUE TITLE
ps_root_recvmsg: Permission denied

### ISSUE THREAD

**Issue by feinerer** [NetworkConfiguration/dhcpcd#237]
I am using dhcpcd 10.0.1 (but also tested with 10.0.2) on OpenBSD 7.3 for IPv6 router solicitation and prefix delegation over PPPoE, almost identical to the setup as described in https://github.com/openbsd/ports/blob/master/net/dhcpcd/pkg/README.

Everything works (i.e., IPv6 connectivity and distributing prefixes) but I have a lot of log messages complaining about:

```
# tail /var/log/messages
Aug  9 06:53:12 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 07:51:20 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 08:45:27 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 09:46:54 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 10:52:03 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 11:51:10 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 12:50:59 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 13:45:53 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 14:51:42 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
Aug  9 15:55:43 fw dhcpcd[57551]: ps_root_recvmsg: Permission denied
```

The function `ps_root_recvmsg` appears in https://github.com/NetworkConfiguration/dhcpcd/blob/master/src/privsep-root.c which is about privilege separation but I have not found out the cause yet.

Anyone else seeing this? Any ideas to fix this?

Thank you!

Best regards,
Ingo

**Comment by rsmarples** [comment]
The error would really come from here: https://github.com/NetworkConfiguration/dhcpcd/blob/1e1268d2de92f148913fc08c1f6efb5b716ac5a1/src/privsep-root.c#L475

I would remove the #ifdef guard around https://github.com/NetworkConfiguration/dhcpcd/blob/1e1268d2de92f148913fc08c1f6efb5b716ac5a1/src/privsep-root.c#L492 so we can work out which command is failing and we can work it from there.

At least on my OpenBSD 7.3 VM everything works, but only over ethernet and not PPPoE. Not that it should really matter.

**Comment by ArthurPeabody** [comment]
ps_root_recvmsg: Operation not permitted

I started getting this message when loading dhcpcd on the
morning of October 20.  It happens between 'sending Router
Solicitation' and 'reading lease: /var/lib/dhcpcd/usb0.lease'.  I see
no new packages loaded after the previous time dhcpcd loaded without
incident.  This doesn't interfere with making a connection, so it
doesn't matter to me.  dhcpcd runs as root.

dhcpcd 10.0.4, Slackware

**Comment by rsmarples** [comment]
That's an interesting data-point which shows it's not OpenBSD specific.

@ArthurPeabody are you able to remove the `#ifdef PRIVSEP_DEBUG` guard as noted above so we can work out which call is at fault please?

**Comment by ArthurPeabody** [comment]
main: control_open: Connection refused
dhcpcd-10.0.4 starting
chrooting as dhcpcd to /var/lib/dhcpcd
Bad system call

Oct 23 11:37:45 [15079]: dhcpcd-10.0.4 starting
Oct 23 11:37:45 [15079]: chrooting as dhcpcd to /var/lib/dhcpcd
Oct 23 11:37:45 [15081]: spawned manager process on PID 15081
Oct 23 11:37:45 [15082]: udev: starting
Oct 23 11:37:45 [15082]: dev: loaded udev
Oct 23 11:37:45 [15081]: spawned privileged proxy on PID 15082
Oct 23 11:37:45 [15082]: ps_root_recvmsgcb: IN cmd 14, psp (nil)
Oct 23 11:37:45 [15082]: ps_root_writeerror: result -1 errno 2
Oct 23 11:37:45 [15081]: spawned network proxy on PID 15083
Oct 23 11:37:45 [15081]: spawned controller proxy on PID 15084
Oct 23 11:37:45 [15081]: network proxy exited unexpectedly from PID 15083, signal=Bad system call

**Comment by rsmarples** [comment]
Wow, that's a different issue!

Please define `SECCOMP_FILTER_DEBUG` in CFLAGS or src/privsep-linux.c so we can work out the failing syscall.
Or enable auditing in the kernel so it will appear in the dmesg.

**Comment by ArthurPeabody** [comment]
Oct 23 15:16:42 [17823]: main: control_open: Connection refused
Oct 23 15:16:42 [17823]: dhcpcd-10.0.4 starting
Oct 23 15:16:42 [17825]: spawned manager process on PID 17825
Oct 23 15:16:42 [17826]: udev: starting
Oct 23 15:16:42 [17826]: dev: loaded udev
Oct 23 15:16:42 [17825]: spawned privileged proxy on PID 17826
Oct 23 15:16:42 [17826]: ps_recvpsmsg: 84
Oct 23 15:16:42 [17825]: spawned network proxy on PID 17827
Oct 23 15:16:42 [17825]: spawned controller proxy on PID 17828
Oct 23 15:16:42 [17825]: network proxy exited unexpectedly from PID 17827, signal=Bad system call
Oct 23 15:16:42 [17823]: chrooting as dhcpcd to /var/lib/dhcpcd
Oct 23 15:16:42 [17826]: ps_recvpsmsg: 0

**Comment by rsmarples** [comment]
That's not telling anything more :/
Can you enable auditing in your kernel? All it should do it put the failing syscall in your dmesg so we can whitelist it in dhcpcd.

**Comment by ArthurPeabody** [comment]
Enabling audit changed nothing in dmesg.

**Comment by ArthurPeabody** [comment]
The error doesn't appear with auditing on.  Weird.

**Comment by ArthurPeabody** [comment]
Hmm... it doesn't appear with auditing off - it's disappeared!  Never mind.

**Comment by rsmarples** [comment]
@ArthurPeabody so there is no more permission error either?

**Comment by ArthurPeabody** [comment]
@rsmarples It's disappeared.  You scared it.  Unfortunately, in the process of building the kernel with auditing enabled, I stumbled onto 2 other unrelated problems.  Bummer.

**Comment by rsmarples** [comment]
I think this issue should now be fixed in dhcpcd-10.0.5
